File: benchmarks/fasta/fasta.rb
IM = 139968
IA = 3877
IC = 29573
LINE_WIDTH = 60

ALU = 'GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA'

class Random
  def initialize(seed)
    @last = seed
  end

  def next
    @last = (@last * IA + IC) % IM
    @last.to_f / IM.to_f
  end
end

def repeat_fasta(seq, n)
  length = seq.length
  pos = 0

  while n > 0
    line_len = [n, LINE_WIDTH].min

    line_len.times do |i|
      print seq[(pos + i) % length]
    end
    puts

    pos = (pos + line_len) % length
    n -= line_len
  end
end

def random_fasta(genelist, n, rng)
  # Build cumulative probabilities
  cum = 0.0
  probs = genelist.map do |_, p|
    cum += p
    cum
  end

  while n > 0
    line_len = [n, LINE_WIDTH].min

    line_len.times do
      r = rng.next
      genelist.each_with_index do |(c, _), k|
        if probs[k] >= r
          print c
          break
        end
      end
    end
    puts

    n -= line_len
  end
end

n = ARGV.size > 0 ? ARGV[0].to_i : 1000

iub = [
  ['a', 0.27],
  ['c', 0.12],
  ['g', 0.12],
  ['t', 0.27],
  ['B', 0.02],
  ['D', 0.02],
  ['H', 0.02],
  ['K', 0.02],
  ['M', 0.02],
  ['N', 0.02],
  ['R', 0.02],
  ['S', 0.02],
  ['V', 0.02],
  ['W', 0.02],
  ['Y', 0.02]
]

homosapiens = [
  ['a', 0.3029549426680],
  ['c', 0.1979883004921],
  ['g', 0.1975473066391],
  ['t', 0.3015094502008]
]

rng = Random.new(42)

puts ">ONE Homo sapiens alu"
repeat_fasta(ALU, 2 * n)

puts ">TWO IUB ambiguity codes"
random_fasta(iub, 3 * n, rng)

puts ">THREE Homo sapiens frequency"
random_fasta(homosapiens, 5 * n, rng)

File: benchmarks/fasta/fasta.rs
use std::io::{self, BufWriter, Write};

const IM: i64 = 139968;
const IA: i64 = 3877;
const IC: i64 = 29573;
const LINE_WIDTH: usize = 60;

const ALU: &[u8] = b"GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

struct Random {
    last: i64,
}

impl Random {
    fn new(seed: i64) -> Self {
        Random { last: seed }
    }

    fn next(&mut self) -> f64 {
        self.last = (self.last * IA + IC) % IM;
        self.last as f64 / IM as f64
    }
}

struct AminoAcid {
    c: u8,
    p: f64,
}

fn repeat_fasta<W: Write>(out: &mut W, seq: &[u8], mut n: usize) -> io::Result<()> {
    let length = seq.len();
    let mut pos = 0;

    while n > 0 {
        let line_len = n.min(LINE_WIDTH);

        for i in 0..line_len {
            out.write_all(&[seq[(pos + i) % length]])?;
        }
        out.write_all(b"\n")?;

        pos = (pos + line_len) % length;
        n -= line_len;
    }
    Ok(())
}

fn random_fasta<W: Write>(
    out: &mut W,
    genelist: &[AminoAcid],
    mut n: usize,
    rng: &mut Random,
) -> io::Result<()> {
    // Build cumulative probabilities
    let mut cum = 0.0;
    let probs: Vec<f64> = genelist
        .iter()
        .map(|aa| {
            cum += aa.p;
            cum
        })
        .collect();

    while n > 0 {
        let line_len = n.min(LINE_WIDTH);

        for _ in 0..line_len {
            let r = rng.next();
            for (k, prob) in probs.iter().enumerate() {
                if *prob >= r {
                    out.write_all(&[genelist[k].c])?;
                    break;
                }
            }
        }
        out.write_all(b"\n")?;

        n -= line_len;
    }
    Ok(())
}

fn main() -> io::Result<()> {
    let n: usize = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(1000);

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());

    let iub = vec![
        AminoAcid { c: b'a', p: 0.27 },
        AminoAcid { c: b'c', p: 0.12 },
        AminoAcid { c: b'g', p: 0.12 },
        AminoAcid { c: b't', p: 0.27 },
        AminoAcid { c: b'B', p: 0.02 },
        AminoAcid { c: b'D', p: 0.02 },
        AminoAcid { c: b'H', p: 0.02 },
        AminoAcid { c: b'K', p: 0.02 },
        AminoAcid { c: b'M', p: 0.02 },
        AminoAcid { c: b'N', p: 0.02 },
        AminoAcid { c: b'R', p: 0.02 },
        AminoAcid { c: b'S', p: 0.02 },
        AminoAcid { c: b'V', p: 0.02 },
        AminoAcid { c: b'W', p: 0.02 },
        AminoAcid { c: b'Y', p: 0.02 },
    ];

    let homosapiens = vec![
        AminoAcid { c: b'a', p: 0.3029549426680 },
        AminoAcid { c: b'c', p: 0.1979883004921 },
        AminoAcid { c: b'g', p: 0.1975473066391 },
        AminoAcid { c: b't', p: 0.3015094502008 },
    ];

    let mut rng = Random::new(42);

    out.write_all(b">ONE Homo sapiens alu\n")?;
    repeat_fasta(&mut out, ALU, 2 * n)?;

    out.write_all(b">TWO IUB ambiguity codes\n")?;
    random_fasta(&mut out, &iub, 3 * n, &mut rng)?;

    out.write_all(b">THREE Homo sapiens frequency\n")?;
    random_fasta(&mut out, &homosapiens, 5 * n, &mut rng)?;

    Ok(())
}
